spark-operator-api/src/test/java/org/apache/spark/k8s/operator/spec/RestartPolicyTest.java

```diff
@@ -41,27 +41,48 @@
 class RestartPolicyTest {
 
   @Test
-  void testAttemptRestartOnState() {
+  void testAlways() {
     for (ApplicationStateSummary stateSummary : ApplicationStateSummary.values()) {
       assertTrue(attemptRestartOnState(Always, stateSummary));
+    }
+  }
+
+  @Test
+  void testNever() {
+    for (ApplicationStateSummary stateSummary : ApplicationStateSummary.values()) {
       assertFalse(attemptRestartOnState(Never, stateSummary));
+    }
+  }
+
+  @Test
+  void testOnFailure() {
+    for (ApplicationStateSummary stateSummary : ApplicationStateSummary.values()) {
       if (!stateSummary.isStopping()) {
         assertFalse(attemptRestartOnState(OnFailure, stateSummary));
-        assertFalse(attemptRestartOnState(OnInfrastructureFailure, stateSummary));
       }
     }
     assertFalse(attemptRestartOnState(OnFailure, Succeeded));
-    assertFalse(attemptRestartOnState(OnInfrastructureFailure, Succeeded));
     assertTrue(attemptRestartOnState(OnFailure, Failed));
-    assertFalse(attemptRestartOnState(OnInfrastructureFailure, Failed));
     assertTrue(attemptRestartOnState(OnFailure, DriverStartTimedOut));
-    assertTrue(attemptRestartOnState(OnInfrastructureFailure, DriverStartTimedOut));
     assertTrue(attemptRestartOnState(OnFailure, DriverReadyTimedOut));
-    assertFalse(attemptRestartOnState(OnInfrastructureFailure, DriverReadyTimedOut));
     assertTrue(attemptRestartOnState(OnFailure, DriverEvicted));
-    assertFalse(attemptRestartOnState(OnInfrastructureFailure, DriverEvicted));
     assertTrue(attemptRestartOnState(OnFailure, ExecutorsStartTimedOut));
+  }
+
+  @Test
+  void testOnInfrastructureFailure() {
+    assertTrue(attemptRestartOnState(OnInfrastructureFailure, DriverStartTimedOut));
     assertTrue(attemptRestartOnState(OnInfrastructureFailure, ExecutorsStartTimedOut));
     assertTrue(attemptRestartOnState(OnInfrastructureFailure, SchedulingFailure));
```

Comment:
The above three are defined as Infrastructure Failures.

https://github.com/apache/spark-kubernetes-operator/blob/7b3dc1c77ecaf1309127bdb7d494ef81915bec34/spark-operator-api/src/main/java/org/apache/spark/k8s/operator/status/ApplicationStateSummary.java#L102-L103